ARCs/arc-0047.md

```diff
@@ -0,0 +1,169 @@
+---
+arc: 47
+title: Logic Signature Templates
+description: Defining templated logic signatures so wallets can safely sign them.
+author: Joe Polny (@joe-p)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/226
+status: Last Call
+last-call-deadline: 2023-08-15
+type: Standards Track
+category: ARC
+created: 2023-07-17
+---
+
+## Abstract
+This standard allows wallets to sign known logic signatures and clearly tell the user what they are signing.
+
+## Motivation
+Currently, most Algorand wallets do not enable the signing of logic signature programs for the purpose of delegation. The rationale is to prevent users from signing malicious programs, but this limitation also prevents non-malicious delegated logic signatures from being used in the Algorand ecosystem. As such, there needs to be a way to provide a safe way for wallets to sign logic signatures without putting users at risk.
+
+## Specification
+
+A logic signature **MUST** be described via the following JSON interface(s):
+
+### Interface
+
+```typescript
+interface LogicSignatureDescription {
+    name: string,
+    description: string,
+    program: string,
+    variables: {
+        variable: string,
+        name: string,
+        type: string,
+        description: string
+    }[]
+}
+```
+
+| Key | Description |
+| --- | ----------- |
+| `name` | The name of the logic signature. **SHOULD** be short and descriptive |
+| `description` | A description of what the logic signature does |
+| `program` | base64 encoding of the TEAL program source
+| `variables` | An array of variables in the program |
+| `variables.variable` | The name of the variable in the templated program. |
+| `variables.name` | Human-friendly name for the variable. **SHOULD** be short and descriptive |
+| `variables.type` | **MUST** be a type defined below in the `type` section |
+| `variables.description` | A description of how this variable is used in the program |
+
+### Variables
+
+A variable in the program **MUST** be start with `TMPL_`
+
+#### Types
+
+Any valid ABI type encoded in base16 (without the leading `0x`) with the following exceptions:
+
+| Type | Description |
+| ---- | ----------- |
+| `address` | 58-character Algorand public address. Typically to be used as an argument to the `addr` opcode. Front-ends **SHOULD** provide a link to the address on an explorer
+| `application` | Application ID. Alias for `uint64`. Front-ends **SHOULD** provide a link to the app on an explorer |
+| `asset` | Asset ID. Alias for `uint64`. Front-ends **SHOULD** provide a link to the asset on an explorer |
+| `string` | ASCII string. Typically used as an argument to `byte`, `method`, or a branching opcode.
```

Comment:
Why ASCII instead of unicode?

Reply:
Mixed up my encodings. Updated in [b02fc9c](https://github.com/algorandfoundation/ARCs/pull/226/commits/b02fc9c9fc267f6d61e764d73bd212e910172403)